src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,165 @@
+        inputs.angleMotorSupplyCurrentOverTime = angleSupplyChargeUsedCoulomb.get();
+        angleStatorChargeUsedCoulomb.update(inputs.angleMotorStatorCurrent);
+        inputs.angleMotorStatorCurrentOverTime = angleStatorChargeUsedCoulomb.get();
+        inputs.angleMotorPosition = angleMotor.getRotorPosition().getValue();
+        inputs.angleMotorVelocity = angleMotor.getVelocity().getValue();
+
+        inputs.angle = getAngle();
+        currentAngle = inputs.angle;
+
+        inputs.angleSetpoint = angleSetpoint;
+
+        inputs.moduleDistance = getModulePosition().distanceMeters;
+        inputs.moduleState = getModuleState();
+    }
+
+    @Override
+    public Rotation2d getAngle() {
+        return new Rotation2d(AngleUtil.normalize(Units.rotationsToRadians(angleMotor.getPosition().getValue())));
+    }
+
+    @Override
+    public void setAngle(Rotation2d angle) {
+        angleSetpoint = angle;
+        Rotation2d error = angle.minus(currentAngle);
+        angleControlRequest.withPosition(angleMotor.getPosition().getValue() + error.getRotations());
+        angleMotor.setControl(angleControlRequest);
+    }
+
+    @Override
+    public double getVelocity() {
+        return driveMotor.getVelocity().getValue();
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        var angleError = angleSetpoint.minus(currentAngle);
+        velocity *= angleError.getCos();
+
+        driveMotorVelocitySetpoint = velocity;
+
+        velocityControlRequest.withVelocity(utils.units.Units.metersToRotations(
+                        velocity,
+                        SwerveConstants.WHEEL_DIAMETER / 2));
+        driveMotor.setControl(velocityControlRequest);
+    }
+
+    @Override
+    public SwerveModuleState getModuleState() {
+        return new SwerveModuleState(
+                utils.units.Units.rpsToMetersPerSecond(
+                        getVelocity(), SwerveConstants.WHEEL_DIAMETER/2)
+                , getAngle()
+        );
+    }
+
+    @Override
+    public SwerveModulePosition getModulePosition() {
+        return new SwerveModulePosition(
+                utils.units.Units.rpsToMetersPerSecond(
+                        driveMotor.getPosition().getValue(),
+                        SwerveConstants.WHEEL_DIAMETER / 2
+                ),
+                getAngle()
+        );
+    }
+
+    @Override
+    public void updateOffset(double offset) {
+        angleMotor.setPosition(
+                ((encoder.getAbsolutePosition() - offset) * Constants.FALCON_TICKS) / SwerveConstants.ANGLE_REDUCTION);
+    }
+
+    @Override
+    public void stopMotor() {
+        driveMotor.stopMotor();
+        angleMotor.stopMotor();
+    }
+
+    @Override
+    public boolean encoderConnected() {
+        return encoder.isConnected();
+    }
+
+    @Override
+    public Command checkModule() {
+        return new RunCommand(() -> {
+            driveMotor.set(0.8);
+            angleMotor.set(0.2);
+        });
+    }
```

Comment:
Do you mean to use it in the `SwerveDrive` class where the function is used?
The `RunCommand` also has a `requirements` parameter but I couldn't get it to work.

Reply:
Ohh I forgot this is not the subsystem code. Your suggestion is good, however, could this method be moved entirely to the subsystem? Remember, all logic should be handled there.

Reply:
I can move the function to the subsystem but I'll still need a function that sets the motors power in the IO.

It says it can't resolve a `<lambda expression>` and `ModuleIOReal` as parameters of the constructor.

Reply:
1. Yep, that's the way to do that. All of the logic should be in the subsystem – we want the same code for real and simulation implementations. The IOs should handle "low-level" access to hardware – setting outputs, getting encoder measurements, etc. I would check though if it's convenient for this IO method to get a `SwerveModuleState`, that way we'd only have one method that sets both speed and angle.
2. The error you get is because the second parameter should be of type `Subsystem`, but a module is not a subsystem. After moving the command to the subsystem, it'll work.

Reply:
> I would check though if it's convenient for this IO method to get a SwerveModuleState

It won't work, the method checks the spin of the motors, a module state gives a specific angle so the angle motor will only spin a little bit.

Reply:
Not sure what you mean

Reply:
When we check a module we want to see both motors spin well, the module state is fine for the drive motor because it will get a velocity control request and spin until we stop the command but the angle motor will get a position control request because he module state has a specific angle so it will reach the setpoint and stop.

Reply:
Oh, I see. So it's fine to have a "raw" method for setting speed, I recommend still to use the `SwerveModuleState` method as much as possible.